Sources/MapboxCoreNavigation/BillingHandler.swift

```diff
@@ -0,0 +1,400 @@
+        /// Indicates that there is no active billing session.
+        case stopped
+        /// There is an active paused billing session.
+        case paused
+        /// There is an active running billing session.
+        case running
+    }
+
+    /// Supported session types.
+    enum SessionType: Equatable {
+        case freeDrive
+        case activeGuidance
+
+        var maxSessionInterval: TimeInterval {
+            switch self {
+            case .activeGuidance:
+                return 43200 /*12h*/
+            case .freeDrive:
+                return 3600 /*1h*/
+            }
+        }
+    }
+
+    /// Shared billing handler instance. There is no other instances of `BillingHandler`.
+    private(set) static var shared: BillingHandler = {
+        let accessToken = NavigationSettings.shared.directions.credentials.accessToken
+        precondition(accessToken != nil, "A Mapbox access token is required. Go to <https://account.mapbox.com/access-tokens/>. In Info.plist, set the MBXAccessToken key to your access token.")
+        let service = ProductionBillingService(accessToken: accessToken ?? "",
+                                               userAgent: URLSession.userAgent)
+        return .init(service: service)
+    }()
+
+    /// The billing service which is used to send billing events.
+    private let billingService: BillingService
+
+    /**
+     A lock which serializes access to variables with underscore: `_sessions` etc.
+     As a convention, all class-level identifiers that starts with `_` should be executed with locked `lock`.
+     */
+    private let lock: NSLock = .init()
+
+    /// All currently active sessions. Running or paused. When session is stopped, it is removed from this variable.
+    private var _sessions: [UUID: Session] = [:]
+
+    private let logger: OSLog = .init(subsystem: "com.mapbox.navigation", category: "Billing")
+
+    /**
+     The state of the billing session.
+
+     - important: This variable is safe to use from any thread.
+     - parameter uuid: Session UUID which is provided in `BillingHandler.beginBillingSession(for:uuid:)`.
+     */
+    func sessionState(uuid: UUID) -> SessionState {
+        lock.lock(); defer {
+            lock.unlock()
+        }
+
+        guard let session = _sessions[uuid] else {
+            return .stopped
+        }
+
+        if session.isPaused {
+            return .paused
+        }
+        else {
+            return .running
+        }
+    }
+
+    /// The token to use for service requests like `Directions` etc. 
+    var serviceSkuToken: String {
+        let sessionTypes: [BillingHandler.SessionType] = [.activeGuidance, .freeDrive]
+
+        for sessionType in sessionTypes {
+            switch billingService.getSessionStatus(for: sessionType) {
+            case .running:
+                return billingService.getSKUTokenIfValid(for: sessionType)
+            case .paused, .stopped:
+                continue
```

Comment:
NIT: It's a bit confusing to use `continue` in the `switch` statement. But not sure if `break` here is less confusing.

Reply:
`break` will stop the loop which isn't a desired outcome.

Reply:
You are right, confused with `if`. 
Anyway, don't see a problem with continue. It is consistent with `return` above.